File: packages/protocol/src/did.ts
import { base58 } from "@scure/base";

const ED25519_CODEC_PREFIX = new Uint8Array([0xed, 0x01]);

/** did:key v0 — Ed25519 public key only (D022). */
export function publicKeyToDid(publicKey: Uint8Array): string {
  if (publicKey.length !== 32) {
    throw new Error("Ed25519 public key must be 32 bytes");
  }
  const prefixed = new Uint8Array(ED25519_CODEC_PREFIX.length + publicKey.length);
  prefixed.set(ED25519_CODEC_PREFIX);
  prefixed.set(publicKey, ED25519_CODEC_PREFIX.length);
  return `did:key:z${base58.encode(prefixed)}`;
}

export function didToPublicKey(did: string): Uint8Array {
  if (!did.startsWith("did:key:")) {
    throw new Error(`Unsupported DID method (v1 supports did:key only): ${did}`);
  }
  const multibase = did.slice("did:key:".length);
  const encoded = multibase.startsWith("z") ? multibase.slice(1) : multibase;
  const decoded = base58.decode(encoded);
  if (
    decoded.length !== 34 ||
    decoded[0] !== ED25519_CODEC_PREFIX[0] ||
    decoded[1] !== ED25519_CODEC_PREFIX[1]
  ) {
    throw new Error("did:key does not contain an Ed25519 public key");
  }
  return decoded.slice(2);
}

export function isDidKey(did: string): boolean {
  try {
    didToPublicKey(did);
    return true;
  } catch {
    return false;
  }
}

File: spec/second-impl/atom_gov/did.py
"""did:key Ed25519 public-key recovery (multibase base58btc)."""

from __future__ import annotations

_ALPHABET = b"123456789ABCDEFGHJKLMNPQRSTUVWXYZabcdefghijkmnopqrstuvwxyz"
_ED25519_PREFIX = bytes([0xED, 0x01])


def _b58decode(data: str) -> bytes:
    if not data:
        raise ValueError("empty base58")
    n = 0
    for ch in data.encode("ascii"):
        try:
            digit = _ALPHABET.index(ch)
        except ValueError as exc:
            raise ValueError("invalid base58 character") from exc
        n = n * 58 + digit
    # Preserve leading zeros (encoded as leading '1's).
    pad = 0
    for ch in data:
        if ch == "1":
            pad += 1
        else:
            break
    raw = n.to_bytes((n.bit_length() + 7) // 8 or 1, "big") if n else b""
    return b"\x00" * pad + raw


def did_to_public_key(did: str) -> bytes:
    if not did.startswith("did:key:"):
        raise ValueError(f"Unsupported DID method: {did}")
    multibase = did[len("did:key:") :]
    encoded = multibase.removeprefix("z")
    decoded = _b58decode(encoded)
    if len(decoded) != 34 or decoded[0:2] != _ED25519_PREFIX:
        raise ValueError("did:key does not contain an Ed25519 public key")
    return decoded[2:]


def is_did_key(did: str) -> bool:
    try:
        did_to_public_key(did)
        return True
    except (ValueError, UnicodeEncodeError):
        return False
